crates/apollo-compiler/src/validation/mod.rs

```diff
@@ -39,6 +44,230 @@ pub trait ValidationDatabase:
+    fn check_union_type_definition(&self, object_type: hir::UnionTypeDefinition) -> Vec<ApolloDiagnostic>;
+    fn check_enum_type_definition(&self, object_type: hir::EnumTypeDefinition) -> Vec<ApolloDiagnostic>;
+    fn check_input_object_type_definition(&self, object_type: hir::InputObjectTypeDefinition) -> Vec<ApolloDiagnostic>;
+    fn check_schema_definition(&self, object_type: hir::SchemaDefinition) -> Vec<ApolloDiagnostic>;
+    fn check_selection_set(&self, selection_set: hir::SelectionSet) -> Vec<ApolloDiagnostic>;
+    fn check_arguments_definition(&self, arguments_def: hir::ArgumentsDefinition) -> Vec<ApolloDiagnostic>;
+    fn check_field_definition(&self, field: hir::FieldDefinition) -> Vec<ApolloDiagnostic>;
+    fn check_input_values(&self, input_values: Arc<Vec<hir::InputValueDefinition>>) -> Vec<ApolloDiagnostic>;
+    fn check_db_definitions(&self, definitions: Arc<Vec<hir::Definition>>) -> Vec<ApolloDiagnostic>;
+    fn check_directive(&self, schema: hir::Directive) -> Vec<ApolloDiagnostic>;
+    fn check_arguments(&self, schema: Vec<hir::Argument>) -> Vec<ApolloDiagnostic>;
+    fn check_field(&self, field: Arc<hir::Field>) -> Vec<ApolloDiagnostic>;
+}
+
+pub fn check_directive_definition(db: &dyn ValidationDatabase, directive: hir::DirectiveDefinition) -> Vec<ApolloDiagnostic> {
+    let mut diagnostics = Vec::new();
+
+    diagnostics.extend(db.check_arguments_definition(directive.arguments));
+
+    diagnostics
+}
+
+pub fn check_object_type_definition(db: &dyn ValidationDatabase, object_type: hir::ObjectTypeDefinition) -> Vec<ApolloDiagnostic> {
+    let mut diagnostics = Vec::new();
+
+    for field in object_type.fields_definition() {
+        diagnostics.extend(db.check_field_definition(field.clone()));
+    }
+
+    diagnostics
+}
+
+pub fn check_interface_type_definition(db: &dyn ValidationDatabase, interface_type: hir::InterfaceTypeDefinition) -> Vec<ApolloDiagnostic> {
+    let mut diagnostics = Vec::new();
+
+    for field in interface_type.fields_definition() {
+        diagnostics.extend(db.check_field_definition(field.clone()));
+    }
+
+    diagnostics
+}
+
+pub fn check_union_type_definition(_db: &dyn ValidationDatabase, _union_type: hir::UnionTypeDefinition) -> Vec<ApolloDiagnostic> {
+    vec![]
+}
+
+pub fn check_enum_type_definition(_db: &dyn ValidationDatabase, _enum_type: hir::EnumTypeDefinition) -> Vec<ApolloDiagnostic> {
+    vec![]
+}
+
+pub fn check_input_object_type_definition(_db: &dyn ValidationDatabase, _input_object_type: hir::InputObjectTypeDefinition) -> Vec<ApolloDiagnostic> {
+    // Not checking the `input_values` here as those are checked as fields elsewhere.
+    vec![]
+}
+
+pub fn check_schema_definition(db: &dyn ValidationDatabase, schema_def: hir::SchemaDefinition) -> Vec<ApolloDiagnostic> {
+    let mut diagnostics = Vec::new();
+
+    for directive in schema_def.directives() {
+        diagnostics.extend(db.check_directive(directive.clone()));
+    }
+
+    diagnostics
+}
```

Comment:
i like this, i think this should work! any individual parts (arguments, directives, fields) can be separated out and validated individually in the future.